Intersect.Client/Interface/Game/Typewriting/Typewriter.cs

```diff
@@ -0,0 +1,145 @@
+        private static List<char> _partialstopChars => ClientConfiguration.Instance.TypewriterPauseCharacters;
+        private static long _partialStopSpeed => ClientConfiguration.Instance.TypewriterPartialStopDelay;
+        private static long _typingSpeed => ClientConfiguration.Instance.TypewriterPartDelay;
+
+        private List<Label> _labels;
+        private string[] _lines;
+        private Label _currentLabel => _labels.ElementAtOrDefault(_lineIndex);
+        private string _currentLine => _lines.ElementAtOrDefault(_lineIndex);
+        private int _lineIndex;
+        private int _charIndex;
+        private char? _lastChar;
+        private long _nextUpdateTime;
+
+        public bool Done { get; private set; }
+
+        public long DoneAt { get; private set; }
+
+        public void NewLine()
+        {
+            _lineIndex++;
+            if (_lineIndex >= _lines.Length)
+            {
+                End();
+                return;
+            }
+            _charIndex = 0;
+            _lastChar = null;
+        }
+
+        public void Initialize(List<Label> labels)
+        {
+            _nextUpdateTime = Timing.Global.Milliseconds;
+            _labels = labels;
+            _lines = _labels.Select(l => l.Text).ToArray();
+            _labels.ForEach(l => l.SetText(string.Empty)); // Clear the text from the labels. The writer is the captain now
+            _lineIndex = 0;
+            _charIndex = 0;
+            _lastChar = null;
+            Done = false;
+        }
+
+        public void Write(string voice)
+        {
+            if (Done)
+            {
+                return;
+            }
+
+            if (string.IsNullOrWhiteSpace(_currentLine) || _currentLabel == default)
+            {
+                End();
+                return;
+            }
+
+            if (Timing.Global.Milliseconds < _nextUpdateTime)
+            {
+                return;
+            }
+
+            if (!string.IsNullOrEmpty(voice) && _charIndex % ClientConfiguration.Instance.TypewriterSoundFrequency == 0)
+            {
+                Audio.StopAllGameSoundsOf(ClientConfiguration.Instance.TypewriterSounds.ToArray());
+                Audio.AddGameSound(voice ?? string.Empty, false);
+            }
+
+            _charIndex++;
+            if (_charIndex >= _currentLine.Length)
+            {
+                _currentLabel?.SetText(_currentLine);
+                NewLine();
+                return;
+            }
+
+            _lastChar = _currentLine[_charIndex - 1];
+
+            var written = _currentLine?.Substring(0, _charIndex) ?? string.Empty;
+            _currentLabel?.SetText(written);
+
+            _nextUpdateTime = Timing.Global.Milliseconds + GetTypingDelayFor(_currentLine.ElementAtOrDefault(_charIndex), _lastChar.Value);
+        }
+
+        private static long GetTypingDelayFor(char currentChar, char? lastChar)
```

Comment:
The method GetTypingDelayFor could be simplified by removing the GetValueOrDefault call and passing the value directly, like `private static long GetTypingDelayFor(char currentChar, char? lastChar = null)`

Reply:
In this case, `GetValueOrDefault` is used because `LastChar` _can_ be null, _but_ we compare it against a couple things, namely `_fullStopChars` and `_partialStopChars` which are lists of _non-nullable_ chars. I find it easier to leave it as a nullable param and just use C# optionals to get the value for comparison to non-null types later, but would have to hear from others as well.